### PR TITLE
Fixes Ametek7270

### PR DESCRIPTION
**Pull request by seb5g**
I'm using this instrument right now and noticed many failures in the code related to:

* not asserting the return value when writing values (dne now by subclassing Instruments with a dedicated check_errors method)
* using right command to translate from index value to float values (Timeconstant and sensitivity)
* sensitivity values are dynamically set depending the IMODE (voltage/current)
* adding other controls/measurements

### CODE REVIEW

File: pymeasure/instruments/ametek/ametek7270.py [hunk 1]
```diff
@@ -30,6 +31,15 @@
 log.addHandler(logging.NullHandler())
 
 
+class Instrument(Instrument):
+
+    def check_errors(self):
+        errors = []
+        if self.read() != '':
+            errors.append(f'Incorrect return from previously set property')
+        return errors
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:40]: As you do not do any work with the lists, I'd do the logic differently:
```suggestion
        if self.read() == '':
            return []
        else:
            return ['Incorrect return from previously set property']
```

File: pymeasure/instruments/ametek/ametek7270.py [hunk 2]
```diff
@@ -50,31 +66,36 @@ class Ametek7270(Instrument):
...
+        "SEN", "SEN %d",
         """ A floating point property that controls the sensitivity
         range in Volts, which can take discrete values from 2 nV to
         1 V. This property can be set. """,
         validator=truncated_discrete_set,
         values=SENSITIVITIES,
-        map_values=True
+        map_values=True,
+        check_set_errors=True,
+        dynamic=True
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:77]:
```suggestion
        dynamic=True,

```
It's good practice to finish even the last line with a comma. That way, you do not have to touch lines (see above `map_values=True`, while adding/removing unrelated lines.

File: pymeasure/instruments/ametek/ametek7270.py [hunk 3]
```diff
@@ -184,16 +218,25 @@ def __init__(self, adapter, name="Ametek DSP 7270", **kwargs):
...
+        self.sensitivity_values = self.SENSITIVITIES_IMODE[0]
 
     def set_differential_mode(self, lineFiltering=True):
         """ Sets instrument to differential mode -- assuming it is in voltage mode """
-        self.write("VMODE 3")
-        self.write("LF %d 0" % 3 if lineFiltering else 0)
+        self.ask("VMODE 3")
+        self.ask("LF %d 0" % 3 if lineFiltering else 0)
+
+    def set_current_mode(self, low_noise=False):
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:229]: Please add a docstring describing, what the method does.

**Comment by bmoneke** [comment]
@seb5g thanks for your contribution. It's good if people help keeping the code base clean.
You might add yourself to the Authors.txt.

Would you mind adding a few protocol tests? They help us maintainers to ensure, that the code continues to work as expected without having an actual device.
If you need support writing these tests, don't hesitate to ask.

**Comment by seb5g** [comment]
> Would you mind adding a few protocol tests? They help us maintainers to ensure, that the code continues to work as expected without having an actual device. If you need support writing these tests, don't hesitate to ask.

Hi, I'm not sure to see the point of your protocol tests. As it is (like Mocking the instrument) it looks as if I'm programming twice the communication protocol... I see only one reason to do so, to prevent someone to change the code base as I did just now. But in my case, the code base was not working properly so the tests would not also! It is only by trying with the instrument that i could figure that out, not with the mock tests...

**Comment by bmoneke** [comment]
These mock tests help us to make sure, that the communication generated is the same. That way we can refactor the instrument code or even the pymeasure core code (e.g. `control`).

So yes, they kind of freeze the communication, but that is essential for us maintainers, who cannot test the code with a device.

There are two different development paths of tests:

1. You have a manual with examples. Then you write the tests and you can verify, that you code fulfills these examples.
2. You have a working code and want to make sure, that it continues to work: You write tests for that code. To facilitate test writing, I'm currently working at a test generator (see #882 ), which observes the communication to the device and writes test accordingly. If you have a working code, you can generate the tests easily.

**Comment by seb5g** [comment]
Hello @bmoneke , would you accept this pull request?
Cheers
Seb

File: pymeasure/instruments/ametek/ametek7270.py [hunk 4]
```diff
@@ -50,34 +68,39 @@ class Ametek7270(Instrument):
...
         """ A floating point property that controls the sensitivity
         range in Volts, which can take discrete values from 2 nV to
         1 V. This property can be set. """,
         validator=truncated_discrete_set,
         values=SENSITIVITIES,
-        map_values=True
+        map_values=True,
+        check_set_errors=True,
+        dynamic=True,
+
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:80]: I think that empty line is superfluous
```suggestion

```

File: pymeasure/instruments/ametek/ametek7270.py [hunk 5]
```diff
@@ -50,34 +68,39 @@ class Ametek7270(Instrument):
...
     slope = Instrument.control(
         "SLOPE", "SLOPE %d",
         """ A integer property that controls the filter slope in
         dB/octave, which can take the values 6, 12, 18, or 24 dB/octave.
         This property can be set. """,
         validator=truncated_discrete_set,
         values=[6, 12, 18, 24],
-        map_values=True
+        map_values=True,
+        check_set_errors=True
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:91]: You see, that the line `map_values` got changed, because you add another line, therefore it is good practice, to end every line, even the last one, with a comma:
```suggestion
        map_values=True,
        check_set_errors=True,
```

Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:91]: You see, that the line `map_values` got changed, because you add another line, therefore it is good practice, to end every line, even the last one, with a comma:
```suggestion
        map_values=True,
        check_set_errors=True,
```

File: pymeasure/instruments/ametek/ametek7270.py [hunk 6]
```diff
@@ -172,28 +208,49 @@ class Ametek7270(Instrument):
...
+            self.sensitivity_values = self.SENSITIVITIES_IMODE[1]
 
     def set_channel_A_mode(self):
         """ Sets instrument to channel A mode -- assuming it is in voltage mode """
-        self.write("VMODE 1")
+        self.ask("VMODE 1")
+
+    def identification(self):
+        """Get the instrument ID and firmware version"""
+        return f"{self.id}/{self.ask('VER')}"
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:253]: First I thought, you ment the "id" property, then I notice, that you use that property and add the firmware version.
Why does the ID + firmware version give another form of identification?

File: tests/instruments/ametek/generate_tests.py
Comment (bmoneke) [inline tests/instruments/ametek/generate_tests.py:1]: @bilderbuchi should we add these test files?
I don't think so.

File: tests/instruments/ametek/generate_tests.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    dev = Ametek7270('USB0::0x0A2D::0x001B::16216344::RAW')
+    dev.id
+    dev.identification()
+    dev.shutdown()
+
+
+if __name__ == '__main__':
+    # run()
+    # generate()
+    pass
```
Comment (bmoneke) [inline tests/instruments/ametek/generate_tests.py:108]: What's that for?

**Comment by bmoneke** [comment]
A first round of comments for your code. Normally it takes a few rounds of rework, but that ensures good and uniform style in the codebase. Your code already looks good, so eventually it will enter pymeasure.
Just ping us (with a message or clicking on "request review"), whenever you want us to have another look at the code.

File: pymeasure/instruments/ametek/ametek7270.py [hunk 8]
```diff
@@ -30,6 +31,18 @@
...
+    def check_errors(self):
+        if self.read() == '':
+            return []
+        else:
+            return ['Incorrect return from previously set property']
+
+    def ask(self, command, query_delay=0):
+        return super().ask(command, query_delay).strip()
+
+
```
[the remark this answers is not in the document]
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:45]: That's a good question... It was intended to highlight the change compared to the standard Instrument class. I moved it into my Ametek instrument class directly.

I added some docstrings into the check_errors and ask reimplemented method explaining why, basically:

        """Usually the properties use the `values` method that adds a strip call,
        however several methods use directly the result from ask to be casted into some other types.
        It should therefore also add the strip here
        """

File: pymeasure/instruments/ametek/ametek7270.py (hunk 4 above)
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:80]: done in next commit

File: pymeasure/instruments/ametek/ametek7270.py (hunk 5 above)
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:91]: done in next commit

Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:91]: done in next commit

File: pymeasure/instruments/ametek/ametek7270.py (hunk 6 above)
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:253]: well it is just a matter of better identification

File: tests/instruments/ametek/generate_tests.py
Comment (seb5g) [reply tests/instruments/ametek/generate_tests.py:1]: it's up to you, now that the test have been generated it doesn't matter

File: tests/instruments/ametek/test_ametek7270.py [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+    ) as inst:
+        assert inst.x2 == ''
+
+
+def test_y2_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'Y2.', b'')],
+    ) as inst:
+        assert inst.y2 == ''
```
[the remark this answers is not in the document]
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:90]: well it depends on the lockin mode (single reference or dual harmonic). In single reference mode, only x, y are read as floats all properties indexed with 1 and 2 are empty string and it is the opposite in dual harmonic mode. 

So how do you expect me to write tests in such a situation?

File: pymeasure/instruments/ametek/ametek7270.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bmoneke) [reply pymeasure/instruments/ametek/ametek7270.py]: Do you expect excess spaces, which you have to strip?

At least the conversion to int or float handles spaces well:
`int("  563 ") == 563`

File: tests/instruments/ametek/generate_tests.py (hunk 7 above)
Comment (bmoneke) [reply tests/instruments/ametek/generate_tests.py:108]: git rm --cached file1.txt
git commit -m "remove file1.txt"

See also http://stackoverflow.com/questions/2047465/ddg#2047477

File: pymeasure/instruments/ametek/ametek7270.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py]: no it actually adds a new line character \n (which is removed with the strip). i tried to add this new line character in the termination character like '\n\x00' but it does not well.

File: tests/instruments/ametek/test_ametek7270.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py:90]: With the tests, I realized the type of the return value, so we are talking about the implementation itself.

I would not expect an empty string as a response value, but 0, None, or nan or that an exception is raised.

As it is an invalid query, an exception, None, or nan can be understood. 

Maybe raising ValueError is the best, as you try to read a property, which is invalid to query right now.

[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py:90]: What would expect, if a property is not defined?
0 could bot be distinguished from a valid response, so 0 is not a good answer

[the remark this answers is not in the document]
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:90]: The thing is (and I tested it on the instrument right now), no error is generated from the instrument only empty response. I mean the call on the property is not invalid it is just undefined.

 I'm not sure I understand, I mean I'm not the one writing the response in the implementation but the driver (instrument) is. If it replies, in certain circumstances, with an empty string, what should I do?? Should I add a check_get_error thing and depending on the mode raise a ValueError, is that what you mean? But then it is not more coherent with the instrument driver documentation...

[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py:90]: These drivers, we are writing, hide some details from the user.
Therefore, we introduce changes, we consider appropriate.
Another instrument, for example, returns the temperature times 10. The python code divides by ten and returns an integer, such that the user gets a temperature in degrees.
So there's no problem deviating from the pure instrument communication.

An empty string is not a good response value for an undefined property.
Imagine a user reads that property and tries to calculate something from it.
Then he will get an obscure error, that you can't add Str to int (for example).

If you raise a meaningful exception, the user will notice immediately, what the problem is.

Maybe the best place is the `read` method.

Can you follow me?

[the remark this answers is not in the document]
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:90]: yeah get it, I'll try to write something!

[the remark this answers is not in the document]
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:90]: I looked into the read reimplementation, but how to know where this read comes from and that the empty string response is wrong (in some cases it is not, for instance after a property has been set). Then I thought about the casting, because a float on an empty string raise a ValueError but I saw you discard this error in the values method:

```
 for i, result in enumerate(results):
            try:
                if cast == bool:
                    # Need to cast to float first since results are usually
                    # strings and bool of a non-empty string is always True
                    results[i] = bool(float(result))
                else:
                    results[i] = cast(result)
            except Exception:
                pass  # Keep as string
``` 

and the ckeck_get_error=True will just also call the check_errors

So I tried another thing using the get_process method:

```


    y = Instrument.measurement("Y.",
                               """ Reads the Y value in Volts """,
                               get_process=check_read_not_empty,
                               )
    x1 = Instrument.measurement("X1.",
                                """ Reads the first harmonic X value in Volts """,
                                get_process=check_read_not_empty,
                                )
```

with `check_read_not_empty` is:

```
def check_read_not_empty(value):
    """Called by some properties to check if the reply is not an empty string
    that would mean the properties is currently invalid (probably because the reference mode
    is on single or dual)"""
    if value == '':
        raise ValueError('Invalid response from measurement call')
    else:
        return value
```

I tested this on the instrument, changing dynamically the reference mode, and it works well. What do you think?

File: pymeasure/instruments/ametek/ametek7270.py (hunk 6 above)
Comment (bmoneke) [reply pymeasure/instruments/ametek/ametek7270.py:261]: I'm still not convinced, that "identification" (the long form of ID) is a good name for that combination.

Do you need the firmware version (which could basically change) as an additional identification to the serial number?

File: tests/instruments/ametek/test_ametek7270.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py:90]: That last code looks good. For reasons like this (or calculating a Value etc.), we have "get_process".

Maybe you could try to make the error message more meaningful, for example hinting at the possibly wrong mode.

It's also good to document in the instrument's docstring this changing between sets of commands.

The idea is, that a user does not have to look at the manual while using your driver.

[the remark this answers is not in the document]
Comment (mcdo0486) [reply tests/instruments/ametek/test_ametek7270.py:90]: > alright now how should I modify the tests on these properties that are raising an error? The generator is not catching the error. Do you have a layout for this? I know about pytest.raises(ValueError) but how to write it with the expected protocol?

You can checkout the DSP Signal Recovery lockin amplifier tests:
https://github.com/pymeasure/pymeasure/blob/master/tests/instruments/signalrecovery/test_dspbase.py

Or the TDK lambda tests in my PR branch here:
https://github.com/mcdo0486/pymeasure/tree/dev/tdk_lambda_40-38_80-65/tests/instruments/tdk

The TDK does something kind of similar in the sense that "OK" gets returned by instrument so I needed to use `check_set_errors`. You can see the expected protocol examples there. 

Hope that helps.

[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py:90]: You can put both with statements into each other:
```python
With pytest.raises():
    With expected_protocol() as inst:
        Pass
```

I'd use a parameterized test (getting all these attributes with "geattr(Inst, Name)") for all the raised exceptions, as they are all the same.

You could make these tests work with returning correct values.


Btw. You said, that the instrument returns a newline char at the end of every message. Why don't you include them in the tests? That way you test, whether your code interprets the messages correctly.

[the remark this answers is not in the document]
Comment (mcdo0486) [reply tests/instruments/ametek/test_ametek7270.py:90]: > alright now how should I modify the tests on these properties that are raising an error? The generator is not catching the error. Do you have a layout for this? I know about pytest.raises(ValueError) but how to write it with the expected protocol?

I tried throwing a PR to your branch, but probably easiest to give the example code here:

```
def test_x1_getter():
    with pytest.raises(ValueError):
        with expected_protocol(
                Ametek7270,
                [(b'X1.', b'')],
        ) as inst:
            assert inst.x1 == ''
```

File: pymeasure/instruments/ametek/ametek7270.py [hunk 10]
```diff
@@ -98,36 +101,45 @@ class Ametek7270(Instrument):
         validator=truncated_discrete_set,
         values=TIME_CONSTANTS,
         map_values=True,
-        check_set_errors=True
+        check_set_errors=True,
     )
 
     x = Instrument.measurement("X.",
-                               """ Reads the X value in Volts """
+                               """ Reads the X value in Volts """,
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:108]: I know, that you did not change the docstring itself, so you do not have to change it, but I want to raise your attention, that that docstring is not according to our coding style.

For your information: A merely gettable property should start with "Measure" or "Get", a settable one with "Set" and a property gettable and settable with "Control".
For more details, see https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings

If you want to, you may improve the docstrings. In that case, I'd kindly ask you to remove the Ametek720 from the `grandfathered_docstring_instruments` in "tests/instruments/test_all_instruments.py" such that docstring tests will be performed on this instrument (I just saw, that it is twice in that list).

You might do that docstring rework in a second PR, if you want, to finish this content rework first.

File: pymeasure/instruments/ametek/ametek7270.py [hunk 11]
```diff
@@ -137,94 +149,127 @@ class Ametek7270(Instrument):
...
-                                cast=str
+                                cast=str,
                                 )
 
     def __init__(self, adapter, name="Ametek DSP 7270",
                  read_termination='\x00',
                  write_termination='\x00',
-                 **kwargs):
+                 **kwargs,
+                 ):
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:235]: In this case, you do not need to add the additional comma after kwargs, as `**kwargs` is normally the last entry, collecting all additional keyword arguments. However, adding the comma and new line, is not wrong either.

File: pymeasure/instruments/ametek/ametek7270.py [hunk 12]
```diff
@@ -137,94 +149,127 @@ class Ametek7270(Instrument):
...
+        """Usually the properties use the `values` method that adds a strip call,
+        however several methods use directly the result from ask to be casted into some other types.
+        It should therefore also add the strip here
+        """
+        return super().ask(command, query_delay).strip()
+
+    def set_reference_mode(self, mode: int = 0):
+        """Set the instrument in Single, Dual or harmonic modes
+        :param mode: the integer specifying the mode: 0 for Single, 1 for Dual harmonic and 2 for Dual
+        reference
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:267]: Small improvements:
```suggestion
        """Set the instrument in Single, Dual or harmonic mode.

        :param mode: the integer specifying the mode: 0 for Single, 1 for Dual harmonic, and 2 for
            Dual reference.

```
Reasons:
- The first line is a sentence in imperative mode, which should end with a period
- Between the summary line (first line) and additional text, should be an empty line.

File: pymeasure/instruments/ametek/ametek7270.py [hunk 13]
```diff
@@ -41,6 +57,11 @@ class Ametek7270(Instrument):
         200.0e-3, 500.0e-3, 1.0
     ]
 
+    SENSITIVITIES_IMODE = {0: SENSITIVITIES,
+                           1: list(np.array(SENSITIVITIES) * 1e-6),
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:61]: Is it worth to import the full numpy package for that list of constants?
Isn't it better to generate that list once and then copy the results into the file?

File: pymeasure/instruments/ametek/ametek7270.py [hunk 14]
```diff
@@ -50,161 +71,242 @@ class Ametek7270(Instrument):
...
+        if self.read() == '':
+            return []
+        else:
+            return ['Incorrect return from previously set property']
+
+    def ask(self, command, query_delay=0):
+        """Usually the properties use the `values` method that adds a strip call,
+        however several methods use directly the result from ask to be casted into some other types.
+        It should therefore also add the strip here
+        """
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:261]:
```suggestion
        """Send a command and read the response, stripping white spaces.

        Usually the properties use the `values` method that adds a strip call,
        however several methods use directly the result from ask to be casted into some other types.
        It should therefore also add the strip here, as all responses end with a newline character.
        """
```
I reformatted the docstring according to https://peps.python.org/pep-0257/

First a summary line, then a more extensive description (here I included, why it is necessary).

**Review by bmoneke — CHANGES_REQUESTED** [review]
Hey @seb5g 
thanks for your reformatting. The code looks good, a few questions remain still open (e.g. making the tests pass) and I added a few more. Mainly on docstring style.

File: pymeasure/instruments/ametek/ametek7270.py (hunk 10 above)
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:108]: Ok I'll see that later

**Comment by seb5g** [comment]
One side question for you guys maintaining Pymeasure:
I'm the main author/developer of PyMoDAQ (pymodaq.cnrs.fr) that among other things has the same purpose as pymeasure (but focused on proposing highly modular GUI for data recording as a function of variable parameters). I think it would be highly  valuable for both our projects to be more aware of what each package do. For this I'd like to invite one of you on the PyMoDAQ's days, a workshop I'm organising next 16/17 october in Paris to present features of pymeasure and discuss of what could be done at the interface of our packages.

Are you interested?

File: pymeasure/instruments/ametek/ametek7270.py (hunk 13 above)
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:61]: Well I guessed numpy was probably used in pymeasure but you're right, I modified this with list comprehension:

```
1: [sen * 1e-6 for sen in SENSITIVITIES],
```

File: pymeasure/instruments/ametek/ametek7270.py (hunk 11 above)
Comment (seb5g) [reply pymeasure/instruments/ametek/ametek7270.py:235]: done

**Comment by seb5g** [comment]
There is something I don't quite understand with tests that should raise a Value Error. I just committed a try but the expected protocol in that case is just returning the value i've put in the protocol. Not raising the ValueError. I feel the ValueError should be in the expected protocol object but not sure one can do that...thanks for your thought on this.

**Comment by mcdo0486** [comment]
> There is something I don't quite understand with tests that should raise a Value Error. I just committed a try but the expected protocol in that case is just returning the value i've put in the protocol. Not raising the ValueError. I feel the ValueError should be in the expected protocol object but not sure one can do that...thanks for your thought on this.

`ValueError` is supposed to happen when the instrument returns `''` right?

So you can change https://github.com/PyMoDAQ/pymeasure/blob/346e57b2daf4a41783f473d236f461c8d474bcb0/tests/instruments/ametek/test_ametek7270.py#L94 to return empty byte string.

```
def test_failing_properties(method, command):
    """in standard single reference mode, these tests should raise a ValueError"""
    with pytest.raises(ValueError):
        with expected_protocol(
                Ametek7270,
                [(f'{command}'.encode(), b'')]
        ) as inst:
           getattr(inst, method) == 0.0
```

File: pymeasure/instruments/ametek/ametek7270.py (hunk 13 above)
Comment (bmoneke) [reply pymeasure/instruments/ametek/ametek7270.py]: Yes, that is even better!

**Comment by bmoneke** [comment]
> One side question for you guys maintaining Pymeasure: I'm the main author/developer of PyMoDAQ (pymodaq.cnrs.fr) that among other things has the same purpose as pymeasure (but focused on proposing highly modular GUI for data recording as a function of variable parameters). I think it would be highly valuable for both our projects to be more aware of what each package do. For this I'd like to invite one of you on the PyMoDAQ's days, a workshop I'm organising next 16/17 october in Paris to present features of pymeasure and discuss of what could be done at the interface of our packages.
> 
> Are you interested?

Oh, interesting project, thanks for telling about that module. It's always good to connect and learn from each other. Thanks for the invitation. I made my fellow maintainers aware of your offer, let's see, if we can manage an attendance.

By the way, we are developing a communication protocol [LECO](https://github.com/pymeasure/leco-protocol) for connecting instrument drivers and data control/acquisition among different processes and computers. Maybe our thought can help you with your remote control option and maybe you can give some feedback into our design process.

**Comment by seb5g** [comment]
hi @bmoneke , I think this PR is good for another review!

**Comment by seb5g** [comment]
Oh and by the way I think most of the functionalities in this driver is redundant with the one is signalrecovery DSP7265 (actually Ametek bought SignalRecovery hence the different name) but i would in fact suggest to move everything in signalrecovery as it is the brand manufacturer and create some DSP base class with everything shared among the two instruments.

**Comment by bmoneke** [comment]
> Oh and by the way I think most of the functionalities in this driver is redundant with the one is signalrecovery DSP7265 (actually Ametek bought SignalRecovery hence the different name) but i would in fact suggest to move everything in signalrecovery as it is the brand manufacturer and create some DSP base class with everything shared among the two instruments.

You can keep them in two distinct manufacturers, if you deem it best to find the device. I'd go for the name printed on the device, as a potential user will search for that brand name. A similar story is with Teledyne (new name) and LeCroy (old name) or hp and keysight.
What we did in these cases, is a link in the manufacturer documentation (e.g. Teledyne says, that LeCroy contains older models and LeCroy states that Teledyne contains newer models).

Shall I review this instrument first and merge it and afterwards you create that common DSP class, or do you want to do the parent DSP class in this PR?

**Comment by mcdo0486** [comment]
> > Oh and by the way I think most of the functionalities in this driver is redundant with the one is signalrecovery DSP7265 (actually Ametek bought SignalRecovery hence the different name) but i would in fact suggest to move everything in signalrecovery as it is the brand manufacturer and create some DSP base class with everything shared among the two instruments.
> 
> You can keep them in two distinct manufacturers, if you deem it best to find the device. I'd go for the name printed on the device, as a potential user will search for that brand name. A similar story is with Teledyne (new name) and LeCroy (old name) or hp and keysight. What we did in these cases, is a link in the manufacturer documentation (e.g. Teledyne says, that LeCroy contains older models and LeCroy states that Teledyne contains newer models).
> 
> Shall I review this instrument first and merge it and afterwards you create that common DSP class, or do you want to do the parent DSP class in this PR?

I also agree you should keep them as separate manufacturers. 

However I think your `Ametek7270` class can inherit the `DSPBase` class from `pymeasure/instruments/signalrecovery/dsp_base.py` to create a new class, like what is going on with `DSP7265` from `pymeasure/instruments/signalrecovery/dsp7265.py`.

I'm not the original driver author but I refactored the driver to a `DSPBase` for use with `DSP7225`. I have access to 7225 model, and I think I can get access to the 7265 model. It would be good to implement any fixes to `DSPBase`, some of the functionality I haven't tested because it only applies to 7265 and I guess 7270.

**Comment by seb5g** [comment]
I'd like you to accept the PR as it is right now. 

And as @mcdo0486 suggested, I'll work later on using the DSPBase that includes almost everything already in the Ametek7270. 

One question on the way for @mcdo0486 : why modifying the read method to replace the Null character and not use it as termination character as I do?

File: pymeasure/instruments/ametek/ametek7270.py [hunk 15]
```diff
@@ -50,161 +71,244 @@ class Ametek7270(Instrument):
...
             adapter,
             name,
-            **kwargs
-        )
+            read_termination=read_termination,
+            write_termination=write_termination,
+            **kwargs)
+
+    def check_errors(self):
+        """mandatory to be used for property setter
```
Comment (bmoneke) [inline pymeasure/instruments/ametek/ametek7270.py:244]: As it is only for the setter, it should be `check_set_errors` (since #883 ):
```suggestion
    def check_set_errors(self):
        """mandatory to be used for property setter
```

File: pymeasure/instruments/ametek/ametek7270.py (hunk 6 above)
Comment (bmoneke) [reply pymeasure/instruments/ametek/ametek7270.py:300]: What do you think about `extended_identification` to state, that it is more than the mere identification, as the serial number is already the identification of the device.

File: tests/instruments/ametek/test_ametek7270.py [hunk 16]
```diff
@@ -0,0 +1,508 @@
...
+            [],
+    ):
+        pass  # Verify the expected communication.
+
+
+@pytest.mark.parametrize("comm_pairs, value", (
+    ([(b'SEN', b'27\n')],
+     1.0),
+    ([(b'SEN', b'27\n')],
+     1.0),
```
Comment (bmoneke) [inline tests/instruments/ametek/test_ametek7270.py:19]: What's the difference between those two cases?

File: tests/instruments/ametek/test_ametek7270.py [hunk 17]
```diff
@@ -0,0 +1,508 @@
...
+    ([(b'REFN', b'7\n')],
+     7.0),
+    ([(b'REFN', b'7\n')],
+     7.0),
+    ([(b'REFN', b'7\n')],
+     7.0),
+    ([(b'REFN', b'7\n')],
+     7.0),
+    ([(b'REFN', b'7\n')],
+     7.0),
```
Comment (bmoneke) [inline tests/instruments/ametek/test_ametek7270.py:111]: Why do you run the same test six times? What do you expect to change between these tests?

File: tests/instruments/ametek/test_ametek7270.py [hunk 18]
```diff
@@ -0,0 +1,508 @@
...
+            [(b'VMODE 1', b'')],
+    ) as inst:
+        assert inst.set_channel_A_mode() is None
+
+
+def test_identification():
+    with expected_protocol(
+            Ametek7270,
+            [(b'ID', b'7270\n'),
+             (b'VER', b'2.11\n')],
```
Comment (bmoneke) [inline tests/instruments/ametek/test_ametek7270.py:498]: Ah, now I understand: `id` does not return the serial number but the device type.

**Review by bmoneke — APPROVED** [review]
A few comments with only little changes, but it looks good in overall.

File: tests/instruments/ametek/test_ametek7270.py (hunk 17 above)
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:111]: well it was autogenerated from your auto code...I'll change that

File: tests/instruments/ametek/test_ametek7270.py (hunk 16 above)
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:19]: well it was autogenerated from your auto code...I'll change that

**Comment by seb5g** [comment]
> > One side question for you guys maintaining Pymeasure: I'm the main author/developer of PyMoDAQ (pymodaq.cnrs.fr) that among other things has the same purpose as pymeasure (but focused on proposing highly modular GUI for data recording as a function of variable parameters). I think it would be highly valuable for both our projects to be more aware of what each package do. For this I'd like to invite one of you on the PyMoDAQ's days, a workshop I'm organising next 16/17 october in Paris to present features of pymeasure and discuss of what could be done at the interface of our packages.
> > Are you interested?
> 
> Oh, interesting project, thanks for telling about that module. It's always good to connect and learn from each other. Thanks for the invitation. I made my fellow maintainers aware of your offer, let's see, if we can manage an attendance.
> 
> By the way, we are developing a communication protocol [LECO](https://github.com/pymeasure/leco-protocol) for connecting instrument drivers and data control/acquisition among different processes and computers. Maybe our thought can help you with your remote control option and maybe you can give some feedback into our design process.

do you think you could reach me by email to discuss about this? My work email address is sebastien.weber@cemes.fr

File: tests/instruments/ametek/test_ametek7270.py (hunk 17 above)
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py]: if you execute one property several times, it accumulates it. The idea is, that you may test different values.

File: tests/instruments/ametek/test_ametek7270.py (hunk 18 above)
Comment (bmoneke) [reply tests/instruments/ametek/test_ametek7270.py:470]: It's not yet in a pymeasure release. Therefore, I think it is acceptable, that the instrument changes a little in behaviour. If someone uses this instrument, they have it in another place or they are on a different branch. So they know, when they switch to a pymeasure release and should do that change consciously.

I feel bad about introducing a deprecated feature into a first release.

**Comment by seb5g** [comment]
I just solved the conflict of authors, so you should be good to go with the merge!

File: tests/instruments/ametek/test_ametek7270.py (hunk 18 above)
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py:470]: Yeah it's fine, people know things are evolving!

File: tests/instruments/ametek/test_ametek7270.py [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+        assert inst.set_channel_A_mode() is None
+
+
+def test_id():
+    with expected_protocol(
+            Ametek7270,
+            [(b'ID', b'7270\n'),
+             (b'VER', b'2.11\n')],
+    ) as inst:
+        assert inst.identification() == '7270/2.11'
```
Comment (bmoneke) [inline tests/instruments/ametek/test_ametek7270.py:472]: You forgot to change that:
```suggestion
        assert inst.id == '7270/2.11'
```

**Review by bmoneke — APPROVED** [review]
It's good except for a failing test, due to not changing identification to id in the test itself.

File: tests/instruments/ametek/test_ametek7270.py (hunk 19 above)
Comment (seb5g) [reply tests/instruments/ametek/test_ametek7270.py]: indeed, you have a keen eye! That's done, thanks for the extended review!

**Comment by bmoneke** [comment]
Thanks for your contribution.
